cryptography/src/bls12381/primitives/poly.rs

```diff
@@ -208,61 +208,63 @@ impl<C: Element> Poly<C> {
     }
 
     /// Recover the polynomial's constant term given at least `t` polynomial evaluations.
-    pub fn recover(t: u32, mut evals: Vec<Eval<C>>) -> Result<C, Error> {
+    ///
+    /// # Warning
+    ///
+    /// This function assumes that each evaluation is unique.
+    pub fn recover<'a, I>(t: u32, evals: I) -> Result<C, Error>
+    where
+        C: 'a,
+        I: IntoIterator<Item = &'a Eval<C>>,
+    {
         // Reference: https://github.com/celo-org/celo-threshold-bls-rs/blob/a714310be76620e10e8797d6637df64011926430/crates/threshold-bls/src/poly.rs#L131-L165
 
-        // Ensure there are enough shares
+        // Ensure we have enough shares
         let t = t as usize;
+        let mut evals = evals.into_iter().collect::<Vec<_>>();
         if evals.len() < t {
-            return Err(Error::InvalidRecovery);
+            return Err(Error::NotEnoughPartialSignatures(t, evals.len()));
         }
 
         // Convert the first `t` sorted shares into scalars
-        let mut err = None;
-        evals.sort_by(|a, b| a.index.cmp(&b.index));
+        evals.sort_by_key(|e| e.index);
```

Comment:
consider not sorting if `evals.len() == t`

Reply:
Will leave this optimization off for now.

Something to reconsider in #719.